opentelemetry-sdk/src/opentelemetry/sdk/_metrics/metric_reader.py

```diff
@@ -12,6 +12,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from abc import ABC, abstractmethod
 
-class MetricReader:
-    pass
+
+class MetricReader(ABC):
```

Comment:
I have some recommendation here based on asking Reiley questions, opening a few bugs in the spec and working on the prototype. Update this `MetricReader` ABC so:

- When the MeterProvider's constructor receives the MetricReaders, it needs to give the MetricReader a way to call [`MeterProvider._measurement_consumer.collect()`](https://github.com/open-telemetry/opentelemetry-python/blob/fb605388dda13933b2242a208e0039518eec01ce/opentelemetry-sdk/src/opentelemetry/sdk/_metrics/measurement_consumer.py#L42). The way I did this in the prototype is with a `MetricReader._set_collect_callback(func: Callable[[], Iterable[Metric]) -> None` function. When a MetricReader is passed to the MeterProvider's constructor, the MeterProvider will do something like this:
  ```py
        # bind readers to collection function
        for metric_reader in self._metric_readers:
            metric_reader._set_collect_callback(
                partial(self._measurement_processor.collect, metric_reader)
            )
  ```
- `MetricReader.collect()` should be marked `@final` and subclasses shouldn't override it. See guidance [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#:~:text=Note%3A%20it%20is%20expected%20that%20the%20MetricReader.Collect%20implementations%20will%20be%20provided%20by%20the%20SDK%2C%20so%20it%20is%20RECOMMENDED%20to%20prevent%20the%20user%20from%20accidentally%20overriding%20it%2C%20if%20possible%20(e.g.%20final%20in%20C%2B%2B%20and%20Java%2C%20sealed%20in%20C%23).) based on https://github.com/open-telemetry/opentelemetry-specification/issues/2139
- Add something like a `@abstracmethod def _receive_metrics(metrics: Iterable[Metric])` which concrete implementations of MetricReader must implement. The final impl of `collect()` will just invoke this function.

So in Periodic Reader, you could basically just do this
  ```py
  def _receive_metrics(metrics: Iterable[Metric]) -> Status:
      self._exporter.export(metrics)
  ```

And add a timer thread to call to `collect()` (see this [guidance](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#:~:text=is%2030000%20(milliseconds).-,One%20possible%20implementation%20of%20periodic%20exporting%20MetricReader%20is%20to%20inherit,task%20calls%20Collect()%20which%20passes%20metrics%20to%20the%20push%20exporter.,-MetricExporter))

Reply:
> I have some recommendation here based on asking Reiley questions, opening a few bugs in the spec and working on the prototype. Update this `MetricReader` ABC so:
> 
> * When the MeterProvider's constructor receives the MetricReaders, it needs to give the MetricReader a way to call [`MeterProvider._measurement_consumer.collect()`](https://github.com/open-telemetry/opentelemetry-python/blob/fb605388dda13933b2242a208e0039518eec01ce/opentelemetry-sdk/src/opentelemetry/sdk/_metrics/measurement_consumer.py#L42). The way I did this in the prototype is with a `MetricReader._set_collect_callback(func: Callable[[], Iterable[Metric]) -> None` function. When a MetricReader is passed to the MeterProvider's constructor, the MeterProvider will do something like this:
>   ```python
>         # bind readers to collection function
>         for metric_reader in self._metric_readers:
>             metric_reader._set_collect_callback(
>                 partial(self._measurement_processor.collect, metric_reader)
>             )
>   ```

Just curious, why don't we just pass the measurement processor to the metric reader?

Reply:
I'm fine with either way. My reason for doing it this way was to hide the MeasurementConsumer from the MetricReader. If we decide to gut the MeasurementConsumer stuff, the MetricReader doesn't need to be updated